lib/Service/Classification/FeatureExtraction/SubjectExtractor.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\Mail\Service\Classification\FeatureExtraction;
+
+use OCA\Mail\Account;
+use OCA\Mail\Db\Message;
+use Rubix\ML\Datasets\Labeled;
+use Rubix\ML\Datasets\Unlabeled;
+use Rubix\ML\Transformers\MultibyteTextNormalizer;
+use Rubix\ML\Transformers\TfIdfTransformer;
+use Rubix\ML\Transformers\WordCountVectorizer;
+use RuntimeException;
+use function array_column;
+use function array_map;
+
+class SubjectExtractor implements IExtractor {
+	private WordCountVectorizer $wordCountVectorizer;
+	private TfIdfTransformer $tfidf;
+	private int $max = -1;
+
+	public function __construct() {
+		// Limit vocabulary to limit memory usage
+		$this->wordCountVectorizer = new WordCountVectorizer(500);
+		$this->tfidf = new TfIdfTransformer();
+	}
+
+	public function getWordCountVectorizer(): WordCountVectorizer {
+		return $this->wordCountVectorizer;
+	}
+
+	public function setWordCountVectorizer(WordCountVectorizer $wordCountVectorizer): void {
+		$this->wordCountVectorizer = $wordCountVectorizer;
+		$this->limitFeatureSize();
+	}
+
+	public function getTfIdf(): TfIdfTransformer {
+		return $this->tfidf;
+	}
+
+	public function setTfidf(TfIdfTransformer $tfidf): void {
+		$this->tfidf = $tfidf;
+	}
+
+	public function prepare(Account $account, array $incomingMailboxes, array $outgoingMailboxes, array $messages): void {
+		/** @var array<array-key, array<string, string>> $data */
+		$data = array_map(static function (Message $message) {
+			return [
+				'text' => $message->getSubject() ?? '',
+				'label' => $message->getFlagImportant() ? 'i' : 'ni',
+			];
+		}, $messages);
+
+		// Fit transformers
+		Labeled::build(
+			array_column($data, 'text'),
+			array_column($data, 'label'),
+		)
```

Comment:
nit³: bracket position is somewhat weird because it looks like this is the end of the statement. might need an indent or the method call to move up :thinking:

Reply:
I'll collapse it into a single line.